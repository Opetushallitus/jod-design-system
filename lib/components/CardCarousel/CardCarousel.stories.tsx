import type { Meta, StoryObj } from '@storybook/react';
import { MediaCard } from '../MediaCard/MediaCard';
import { CardCarousel } from './CardCarousel';

const meta = {
  title: 'Cards/CardCarousel',
  component: CardCarousel,
  tags: ['autodocs'],
} satisfies Meta<typeof CardCarousel>;

export default meta;

type Story = StoryObj<typeof meta>;

const design = {
  type: 'figma',
  url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=7031-3025&t=hppCWVuC76hePJaZ-4',
};

const Link = ({ children, to, className }: { children: React.ReactNode; to?: string | object; className?: string }) => (
  <a href={typeof to === 'string' ? to : '#'} className={className}>
    {children}
  </a>
);

export const Default: Story = {
  decorators: [
    (Story) => (
      <div className="ds:max-w-[1140px] ds:px-5">
        <Story />
      </div>
    ),
  ],
  parameters: {
    design,
    docs: {
      description: {
        story: 'This is a carousel component for displaying MediaCards.',
      },
    },
  },
  args: {
    items: Array.from({ length: 15 }, (_, i) => ({
      id: `card-${i + 1}`,
      component: (
        <MediaCard
          label={`Card ${i + 1}`}
          description={`Description ${i + 1}`}
          imageSrc="https://images.unsplash.com/photo-1523464862212-d6631d073194?q=80&w=260"
          imageAlt={`Image ${i + 1}`}
          to="#"
          linkComponent={Link}
          tags={[
            { label: 'Lorem', to: '#' },
            { label: 'Ipsum', to: '#' },
            { label: 'Dolor', to: '#' },
          ]}
        />
      ),
    })),
    itemWidth: 260,
    translations: {
      nextTrigger: 'Next page',
      prevTrigger: 'Previous page',
      indicator: (idx: number) => `Go to page ${idx + 1}`,
    },
    className: 'ds:p-7 ds:-m-7',
  },
};

export const DifferentCardHeight: Story = {
  decorators: [
    (Story) => (
      <div className="ds:max-w-[1140px] ds:px-5">
        <Story />
      </div>
    ),
  ],
  parameters: {
    design,
    docs: {
      description: {
        story: 'This is a carousel component for displaying MediaCards.',
      },
    },
  },
  args: {
    items: [
      {
        id: `card-1`,
        component: (
          <MediaCard
            label="Lorem ipsum"
            description="Lorem ipsum dolor sit amet, consectetur adipiscing elit."
            imageSrc="https://images.unsplash.com/photo-1523464862212-d6631d073194?q=80&w=260"
            imageAlt="Image 1"
            to="#"
            linkComponent={Link}
            tags={[
              { label: 'lorem', to: '#' },
              { label: 'ipsum', to: '#' },
              { label: 'dolor', to: '#' },
            ]}
          />
        ),
      },
      {
        id: `card-2`,
        component: (
          <MediaCard
            label="Lorem ipsum dolor sit amet"
            description="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat."
            imageSrc="https://images.unsplash.com/photo-1523464862212-d6631d073194?q=80&w=260"
            imageAlt="Image 2"
            to="#"
            linkComponent={Link}
            tags={[
              { label: 'lorem', to: '#' },
              { label: 'ipsum', to: '#' },
            ]}
          />
        ),
      },
      {
        id: `card-3`,
        component: (
          <MediaCard
            label="Lorem ipsum"
            description="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
            imageSrc="https://images.unsplash.com/photo-1523464862212-d6631d073194?q=80&w=260"
            imageAlt="Image 3"
            to="#"
            linkComponent={Link}
            tags={[
              { label: 'lorem', to: '#' },
              { label: 'ipsum', to: '#' },
              { label: 'dolor', to: '#' },
              { label: 'sit', to: '#' },
              { label: 'amet', to: '#' },
            ]}
          />
        ),
      },
    ],
    itemWidth: 260,
    translations: {
      nextTrigger: 'Next page',
      prevTrigger: 'Previous page',
      indicator: (idx: number) => `Go to page ${idx + 1}`,
    },
    className: 'ds:p-7 ds:-m-7',
  },
};
