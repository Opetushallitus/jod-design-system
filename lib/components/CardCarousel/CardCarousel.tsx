import React from 'react';
import { MdChevronLeft, MdChevronRight } from 'react-icons/md';

export interface CardCarouselItem {
  /** Id to be used as key during iteration */
  id: string;
  /** Component to be rendered in the carousel */
  component: React.ReactNode;
}

export interface CardCarouselProps {
  /** Items to show in the carousel */
  items?: CardCarouselItem[];
  /** Width of each item in the carousel */
  itemWidth: number;
  /** Gap between items in the carousel. Default is 16px */
  gap?: number;
  /** Translations for accessibility */
  translations: {
    nextTrigger: string;
    prevTrigger: string;
    indicator: (index: number) => string;
  };
  className?: string;
}
export const CardCarousel = ({ items = [], translations, itemWidth, gap = 16, className = '' }: CardCarouselProps) => {
  const containerRef = React.createRef<HTMLUListElement>();
  const [itemsPerPage, setItemsPerPage] = React.useState(1);
  const [pageNr, setPageNr] = React.useState(0);
  const [pageCount, setPageCount] = React.useState(0);
  const [isFirstPage, setIsFirstPage] = React.useState(false);
  const [isLastPage, setIsLastPage] = React.useState(false);
  const getPageCount = React.useCallback(() => Math.ceil(items.length / itemsPerPage), [itemsPerPage, items.length]);

  React.useEffect(() => {
    const { current } = containerRef;
    if (current) {
      current.scrollTo({ left: pageNr * itemsPerPage * (itemWidth + gap), behavior: 'smooth' });
    }
  }, [itemsPerPage, containerRef, itemWidth, pageNr, gap]);

  const goToNextPage = () => {
    if (!isLastPage) {
      setPageNr(pageNr + 1);
    }
  };
  const goToPreviousPage = () => {
    if (!isFirstPage) {
      setPageNr(pageNr - 1);
    }
  };

  const goToPage = (page: number) => () => {
    setPageNr(page);
  };

  const handleEnterPress = (callback: () => void) => (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      callback();
    }
  };

  React.useEffect(() => {
    setPageCount(getPageCount());
    setIsFirstPage(pageNr === 0);
    setIsLastPage(pageNr === pageCount - 1);
  }, [itemsPerPage, items.length, pageNr, pageCount, getPageCount]);

  React.useEffect(() => {
    const { current } = containerRef;

    const handleResize = (entries: ResizeObserverEntry[]) => {
      for (const entry of entries) {
        if (entry.target == current) {
          setItemsPerPage(Math.max(Math.floor((entry.contentRect.width + gap) / (itemWidth + gap)), 1));
          setPageCount(getPageCount());
          setPageNr(Math.min(Math.max(0, pageNr), getPageCount() - 1));
        }
      }
    };

    const resizeObserver = new ResizeObserver(handleResize);

    if (current) {
      resizeObserver.observe(current);
    }

    return () => {
      if (current) {
        resizeObserver.unobserve(current);
      }
    };
  }, [itemsPerPage, getPageCount, itemWidth, items.length, pageNr, containerRef, gap]);

  return (
    <>
      <ul
        ref={containerRef}
        role="group"
        aria-roledescription="carousel"
        className={`ds:flex ds:flex-row ds:overflow-hidden ${className}`.trim()}
        style={{ gap }}
      >
        {items.map((item, index) => {
          // Change the page according to focused item during tab navigation
          const onFocus = () => {
            const pageWhereFocusedItemIs = Math.floor(index / itemsPerPage);
            const focusedItemIsOutsideCurrentPage = pageWhereFocusedItemIs !== pageNr;

            if (focusedItemIsOutsideCurrentPage) {
              setPageNr(pageWhereFocusedItemIs);
            }
          };

          return (
            <li
              key={item.id}
              aria-roledescription="slide"
              onFocus={onFocus}
              className="ds:flex"
              style={{ width: itemWidth }}
            >
              {item.component}
            </li>
          );
        })}
      </ul>
      <div className="ds:flex ds:flex-row ds:gap-2 ds:justify-between ds:items-center ds:p-3">
        <button
          onClick={goToPreviousPage}
          onKeyDown={handleEnterPress(goToPreviousPage)}
          aria-label={translations.prevTrigger}
          disabled={isFirstPage}
          className="ds:size-8 ds:flex ds:justify-center ds:items-center ds:bg-bg-gray-2 ds:rounded-full ds:cursor-pointer ds:disabled:cursor-not-allowed"
        >
          <MdChevronLeft size={24} className={isFirstPage ? 'ds:text-inactive-gray' : 'ds:text-black'} />
        </button>

        <div className="ds:flex ds:flex-row ds:gap-2 ds:justify-center">
          {Array.from({ length: Math.max(1, pageCount) }, (_, page) => (
            <button
              type="button"
              key={page}
              className={`ds:cursor-pointer ds:rounded-full ds:size-4 ${pageNr === page ? 'ds:bg-accent' : 'ds:bg-[#d4d4d4]'}`}
              disabled={pageNr === page}
              aria-pressed={pageNr === page}
              onClick={goToPage(page)}
              onKeyDown={handleEnterPress(goToPage(page))}
            >
              <span className="ds:sr-only">{translations.indicator(page)}</span>
            </button>
          ))}
        </div>

        <button
          onClick={goToNextPage}
          onKeyDown={handleEnterPress(goToNextPage)}
          aria-label={translations.nextTrigger}
          disabled={isLastPage}
          className="ds:size-8 ds:flex ds:justify-center ds:items-center ds:bg-bg-gray-2 ds:rounded-full ds:cursor-pointer ds:disabled:cursor-not-allowed"
        >
          <MdChevronRight size={24} className={isLastPage ? 'ds:text-inactive-gray' : 'ds:text-black'} />
        </button>
      </div>
    </>
  );
};
