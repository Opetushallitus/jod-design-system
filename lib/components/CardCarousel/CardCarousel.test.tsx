import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { CardCarousel, CardCarouselProps } from './CardCarousel';

const renderComponent = (props: Partial<CardCarouselProps> = {}) => {
  const defaultProps: CardCarouselProps = {
    items: [
      { id: '1', component: <div>Item 1</div> },
      { id: '2', component: <div>Item 2</div> },
      { id: '3', component: <div>Item 3</div> },
    ],
    itemWidth: 200,
    gap: 16,
    translations: {
      nextTrigger: 'Next',
      prevTrigger: 'Previous',
      indicator: (index) => `Page ${index + 1}`,
    },
  };

  return render(<CardCarousel {...defaultProps} {...props} />);
};

describe('CardCarousel', () => {
  it('renders the carousel items', () => {
    const { getByText } = renderComponent();
    expect(getByText('Item 1')).toBeInTheDocument();
    expect(getByText('Item 2')).toBeInTheDocument();
    expect(getByText('Item 3')).toBeInTheDocument();
  });

  it('navigates to the next page', () => {
    const { getByLabelText } = renderComponent();
    const nextButton = getByLabelText('Next');
    fireEvent.click(nextButton);
    expect(nextButton).not.toBeDisabled();
  });

  it('navigates to the previous page', () => {
    const { getByLabelText } = renderComponent();
    const prevButton = getByLabelText('Previous');
    fireEvent.click(prevButton);
    expect(prevButton).toBeDisabled();
  });

  it('disables the next button on the last page', () => {
    const { getByLabelText } = renderComponent();
    const nextButton = getByLabelText('Next');
    fireEvent.click(nextButton);
    fireEvent.click(nextButton);
    expect(nextButton).toBeDisabled();
  });

  it('disables the previous button on the first page', () => {
    const { getByLabelText } = renderComponent();
    const prevButton = getByLabelText('Previous');
    expect(prevButton).toBeDisabled();
  });

  it('navigates to a specific page', () => {
    const { getByText } = renderComponent();
    const pageButton = getByText('Page 2');
    fireEvent.click(pageButton);
    expect(pageButton.parentElement).toHaveClass('ds:bg-accent');
  });

  it('handles enter key press for navigation', () => {
    const { getByLabelText } = renderComponent();
    const nextButton = getByLabelText('Next');
    fireEvent.keyDown(nextButton, { key: 'Enter' });
    fireEvent.keyDown(nextButton, { key: 'Enter' });
    expect(nextButton).toBeDisabled();
  });
});
