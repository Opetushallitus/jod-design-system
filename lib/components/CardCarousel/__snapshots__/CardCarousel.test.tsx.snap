// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardCarousel > renders the carousel items 1`] = `
<ul
  aria-roledescription="carousel"
  class="ds:flex ds:flex-row ds:overflow-hidden ds:p-3 ds:relative ds:-m-3"
  role="group"
  style="gap: 16px;"
>
  <li
    aria-roledescription="slide"
    class="ds:flex"
    style="width: 200px;"
  >
    <div>
      Item 1
    </div>
  </li>
  <li
    aria-roledescription="slide"
    class="ds:flex"
    style="width: 200px;"
  >
    <div>
      Item 2
    </div>
  </li>
  <li
    aria-roledescription="slide"
    class="ds:flex"
    style="width: 200px;"
  >
    <div>
      Item 3
    </div>
  </li>
</ul>
`;
